module/errorhandler/erroraware.go

```diff
@@ -0,0 +1,165 @@
+// "downcasting" to context.Context. Yet, we can still type-assert and recover.
+//
+// ThrowIrrecoverable can be a drop-in replacement anywhere we have a context.Context likely
+// to support Irrecoverables. Note: this is not a method
+func ThrowIrrecoverable(ctx context.Context, err error) {
+	signalerAbleContext, ok := ctx.(irrecoverableSignalerCtxt)
+	if ok {
+		signalerAbleContext.signaler.ThrowIrrecoverable(err)
+	}
+	// Be spectacular on how this does not -but should- handle irrecoverables:
+	log.Fatalf("Irrecoverable error signaler not found for context, please implement! Unhandled irrecoverable error %v", err)
+}
+
+////////////////////////////////////////////////
+// Integrating it w/ ReadyDoneAware & friends //
+////////////////////////////////////////////////
+
+// If we want to do it using interface composition (see module.Component), we will need to build on the existing intrefaces
+// note the irrecoverable management needs to be:
+// - set up before the call to start (if the start itself meets an irrecoverable condition)
+// - not throw / return error itself, except to an enclosing context
+
+type Startable interface {
+	Start(runCtx IrrecoverableSignalerContext) error
+}
+
+// later, in RunComponent, the plumbing happens ...
+type Component interface {
+	Startable
+	module.ReadyDoneAware
+}
+type ComponentFactory func() (Component, error)
+
+func RunComponent(parentCtx IrrecoverableSignalerContext, componentFactory ComponentFactory, handler IrrecoverableHandler) {
+
+	component, err := componentFactory()
+	if err != nil {
+		parentCtx.ThrowIrrecoverable(err) // failure to generate the component
+	}
+
+	// reference to per-run signals for the component
+	var cancel context.CancelFunc
+	var done <-chan struct{}
+	var irrecoverables chan error
+
+	// Tells us:
+	// - how to get started
+	// - how to create a closure to program the handler with a continuation
+	restart := func() {
+		// context used to run the component
+		var runCtx context.Context
+		runCtx, cancel = context.WithCancel(parentCtx)
+
+		// signaler used for irrecoverables
+		var signalingCtx IrrecoverableSignalerContext
+		irrecoverables = make(chan error)
+		signalingCtx = WithIrrecoverableSignal(runCtx, &IrrecoverableSignaler{irrecoverables})
+
+		if err = component.Start(signalingCtx); err != nil {
+			// failed to start component: this should not trigger a restart
+			parentCtx.ThrowIrrecoverable(err)
+		}
+		// Anywhere inside the component, we can use signalingCtx.ThrowIrrecoverable(err), if the types support it
+		// and ThrowIrrecoverable(signalingCtx, err) if not
+
+		// wait for Ready
+		<-component.Ready()
+
+		done = component.Done()
+	}
+
+	restart()
+
+	for {
+		select {
+		case err := <-irrecoverables:
+			// shutdown the component,
+			cancel()
+			// here the handler is programmed with a restart continuation
+			handler.OnIrrecoverable(err, restart)
```

Comment:
We need to wait for the component to finish shutting down here.

Reply:
there isn't really any shutting down to do. if we're here, the component has already called `Goexit` and the goroutine is exiting.

Reply:
Given that we reinstantiate the component on every run, and that you and @peterargue are intent on having a reference to the component in the handler (how?), is this really necessary, or could the handler take care of that?

Reply:
I don't think we actually need a reference to the component in the handler, because the handler is explicitly tied to the component that it is called with in `RunComponent`. So no, we don't need to have a reference to component in handler, but therefore I do think we should wait for component to shutdown here.

Even if we reinstantiate, there *could* be undesirable side effects of having two instances of the same component running at the same time?

Reply:
> Even if we reinstantiate, there could be undesirable side effects of having two instances of the same component running at the same time?

For sure.